File: data-pipeline-libraries-java/data-pipeline-core-java/src/main/java/com/enrichmeai/culvert/dataquality/NumericRange.java
package com.enrichmeai.culvert.dataquality;

import java.util.Objects;

/**
 * A closed numeric range {@code [min, max]} used by {@link DataQualityTransform}
 * to detect out-of-range field values.
 *
 * <p>Both bounds are inclusive. Attach a {@code NumericRange} to a
 * {@link com.enrichmeai.culvert.schema.SchemaField} via
 * {@link com.enrichmeai.culvert.schema.SchemaField#withRange(NumericRange)} to opt
 * that field into range validation. {@link DataQualityTransform} reads bounds
 * directly from the schema — no separate side-map is required.
 *
 * @param min The inclusive lower bound.
 * @param max The inclusive upper bound (must be ≥ {@code min}).
 *
 * @since Sprint 14 / issue #73 (T14.1); schema-grounded T14.7 / issue #100
 */
public record NumericRange(double min, double max) {

    public NumericRange {
        if (max < min) {
            throw new IllegalArgumentException(
                    "max (" + max + ") must be >= min (" + min + ")");
        }
    }

    /** Convenience factory. */
    public static NumericRange of(double min, double max) {
        return new NumericRange(min, max);
    }

    /**
     * Returns {@code true} if {@code value} lies within {@code [min, max]} (inclusive).
     *
     * @param value The numeric value to test.
     */
    public boolean contains(double value) {
        return value >= min && value <= max;
    }
}

File: data-pipeline-libraries/data-pipeline-core/src/data_pipeline_core/dataquality/numeric_range.py
"""NumericRange — closed inclusive numeric range [min, max].

Mirrors ``com.enrichmeai.culvert.dataquality.NumericRange`` (Java
record, Sprint 14 / issue #73; schema-grounded T14.7 / issue #100).

Port: T18.2 / issue #118.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class NumericRange:
    """A closed numeric range ``[min, max]`` (both bounds inclusive).

    Used by :class:`~data_pipeline_core.dataquality.DataQualityTransform`
    to detect ``OUT_OF_RANGE`` field values.  Mirrors the Java record
    ``NumericRange(double min, double max)`` (Java lines 20–42):

    * Both bounds are ``float`` (Python) / ``double`` (Java).
    * ``max`` must be ``>=`` ``min`` — ``ValueError`` otherwise (mirrors
      ``IllegalArgumentException`` at Java line 24).
    * :meth:`contains` mirrors ``NumericRange.contains(double)`` (Java
      line 39).

    Attach a ``NumericRange`` to a :class:`~data_pipeline_core.schema.entity.SchemaField`
    via ``SchemaField(... range=NumericRange.of(lo, hi))`` to opt that
    field into range validation.
    """

    min: float
    max: float

    def __post_init__(self) -> None:
        # Mirrors Java compact constructor (NumericRange.java:22-27).
        if self.max < self.min:
            raise ValueError(
                f"max ({self.max}) must be >= min ({self.min})"
            )

    @classmethod
    def of(cls, min: float, max: float) -> "NumericRange":
        """Convenience factory — mirrors ``NumericRange.of(double, double)``
        (Java line 30)."""
        return cls(min=min, max=max)

    def contains(self, value: float) -> bool:
        """Return ``True`` iff ``value`` lies within ``[min, max]`` inclusive.

        Mirrors ``NumericRange.contains(double)`` (Java lines 39-41).
        """
        return self.min <= value <= self.max
